Backend/bug/coordinate wrapping

spreeni opened the pull request:
Closes #835 

When entering a geometry point with `create_coordinate`, it is now cast to geography type and back to ensure longitude wrapping. 

If a latitude is given outside of the valid range [-90, 90], `create_coordinate` will instead throw a `ValueError`. This was done as PostGIS wraps latitude wackily, e.g. (40, 95) -> (40, 85) instead of (-120, 85).

Added a test case in `test_api.py` and a migration script that does the above mentioned back-and-forth casting.

yasp-pland committed:
Added casting to geography and back again on coordinate creation to ensure coordinate wrapping

yasp-pland committed:
Add latitude overflow error handling and tests for coordinate wrapping

yasp-pland committed:
Created a migration script for coordinate wrapping

spreeni commented:
Note: We still have issues in the Frontend if somebody wants to set their location to e.g. the North or South Pole. The displaced longitude for a latitude of 90 or -90 will be infinitely large. And in general the calculated randomized location and circle points can again have invalid coordinates for a valid starting point. Maybe it makes sense to implement a guard here.

Steps to reproduce: Search for "North Pole" or "South Pole" in MapBox

aapeliv commented:
Just doing the cast seems to work for me? This is from postgres:

```
postgres=# select st_asewkt(st_makepoint(0, 95)::geography::geometry);
NOTICE:  Coordinate values were coerced into range [-180 -90, 180 90] for GEOGRAPHY
       st_asewkt
-----------------------
 SRID=4326;POINT(0 85)
(1 row)
```

I think this is fine: if someone does put "North Pole", at least this will be close? Or am I missing something?

Also, please add full regression tests by just creating a list of pairs of ((inpux_lat, input_lng),(output_lat,output_lng)) and running them all through the database directly to double check the output is correct for a set of inputs!

spreeni commented:
The cast works fine as a guard, but the results for latitude wrapping were a bit counter-intuitive to me. If I would go further North and cross the North Pole, I would expect to at some point be on the other half of the globe in terms of longitude. But it is probably right that this should be enough and close enough to the real location in most situations. I'll remove the check for a ValueError then.

And I can add the regression tests, no problem. Is there a convention where I should insert them? And would you test the SQL query directly or the use of `create_coordinate`? For a direct SQL query I would just add it in `test_db.py`, right?

aapeliv commented:
Yeah, `test_db.py` is good. You need to use `create_coordinate` which creates SQL so you need to run it through the database. Let me know on slack if you have any issues.

yasp-pland committed:
Latitude overflow does not throw Exception anymore

yasp-pland committed:
Added regression test for coordinate wrapping

spreeni committed:
Merge branch 'develop' into backend/bug/coordinate-wrapping

spreeni commented:
Removed the Exception throw and added regression tests. One weird thing that PostGIS does is that it generally won't wrap a longitude of -180, but will wrap it to 180 if there is a latitude overflow.

aapeliv commented on app/backend/src/tests/test_db.py:
```diff
@@ -87,6 +88,40 @@ def test_get_parent_node_at_location(testing_communities):
+
+                assert latitude == lat_expected
+
+                # Weird interaction in PostGIS where lng flips at -180 only when there is latitude overflow
+                if lng == -180 and (lat > 90 or lat < -90):
+                    assert longitude == -lng_expected
+                else:
+                    assert longitude == lng_expected
```
Just make a list of `lat,lng,lat_exp,lng_exp` and test them with

```py
assert lat_exp, lng_exp == get_coordinates(...)
```

Something like this. Having this logic in the test is not good, now `lng_expected` is not really the expected lng.

spreeni replied:
Yeah, thought about this, thought it might be ok with the comment. 

Changed it now to a few fixed coordinates.

aapeliv commented on app/backend/src/couchers/migrations/versions/e264cbf0ca07_coordinate_wrapping_for_points_in_users_.py:
```diff
@@ -0,0 +1,38 @@
+    """
+    op.execute("""
+    UPDATE users
+    SET geom = (ST_SetSRID( geom, 4326)::geography )::geometry
```
Remove extra space in front of `geom`

yasp-pland committed:
Removed additional logic from coord wrapping regression test

yasp-pland committed:
Merge branch 'develop' into backend/bug/coordinate-wrapping

yasp-pland committed:
Created new migration for coordinate wrapping; isort & black

aapeliv commented on app/backend/src/couchers/migrations/versions/f4a49acd8801_coordinate_wrapping_for_users_events_.py:
```diff
@@ -0,0 +1,44 @@
+    """
+    Cast points to geography and back to wrap coordinates
+    """
+    op.execute(
+        """
+    UPDATE users
+    SET geom = (ST_SetSRID(geom, 4326)::geography)::geometry
+    """
+    )
```
You can make this single lines and it'll format them much nicer, but not critical

spreeni replied:
True, changed this.

aapeliv reviewed (dismissed):
LGTM!

yasp-pland committed:
Beautified SQL-statements
